File: src/achievements/StackingTemplates.ts
/**
 * stackingTemplates module contains the templates definitions for the achievements
 */

import { constants } from './constants';
import {StackingAchievementTemplate} from './Achievement';

export namespace StackingTemplates {
  //////////////////////// CRITERIA FUNCTIONS ////////////////////////
  export const STANDARD_EASY_CRITERIA_FUNCTION = (x: number) => 10 ** x;
  export const STANDARD_MEDIUM_CRITERIA_FUNCTION = (x: number) => 2 ** ((x + 2) >> 1) * 5 ** ((x + 1) >> 1);
  export const STANDARD_HARD_CRITERIA_FUNCTION = (x: number) => 2 ** (x >> 1) * 5 ** ((x + 1) >> 1);
  export const STANDARD_INFERNAL_CRITERIA_FUNCTION = (x: number) => (2 ** Math.max(0, (x - 1) >> 1)) * (3 ** (x === 1 ? 1 : 0)) * (5 ** (x >> 1));
  export const STANDARD_POINTS_FUNCTION = STANDARD_INFERNAL_CRITERIA_FUNCTION;

  export namespace files {
    //////////////////////// FILES STACKING TEMPLATES ////////////////////////
    export const filesCreatedTemplate = (): StackingAchievementTemplate => ({
      name: 'Creator %d',
      iconDir: 'creator',
      category: constants.category.FILES,
      group: 'Files Created',
      labels: [constants.category.FILES, constants.labels.FILES_CREATED],
      criterias: [constants.criteria.FILES_CREATED],
      criteriasFunctions: [STANDARD_MEDIUM_CRITERIA_FUNCTION],
      description: `Create ${constants.criteria.FILES_CREATED} files`,
      minTier: 0,
      maxTier: 15,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const filesCreatedLanguageTemplates = (): StackingAchievementTemplate[] =>
      constants.labels.LANGUAGES.map(language => ({
        name: `LANGUAGE Creator %d`.replace('LANGUAGE', language),
        iconDir: `%s_creator`.replace('%s', language),
        category: constants.category.FILES,
        group: 'Files Created',
        labels: [constants.category.FILES, constants.labels.FILES_CREATED],
        criterias: [constants.criteria.FILES_CREATED_LANGUAGE.replace('%s', language)],
        criteriasFunctions: [STANDARD_HARD_CRITERIA_FUNCTION],
        description: `Create FILES files in LANGUAGE`.replace('LANGUAGE', language).replace('FILES', constants.criteria.FILES_CREATED_LANGUAGE.replace('%s', language)),
        minTier: 0,
        maxTier: 11,
        pointsFunction: STANDARD_POINTS_FUNCTION,
        hidden: false,
        requires: [],
      }));

    export const filesDeletedTemplate = (): StackingAchievementTemplate => ({
      name: 'Deleter %d',
      iconDir: 'deleter',
      category: constants.category.FILES,
      group: 'Files Deleted',
      labels: [constants.category.FILES, constants.labels.FILES_DELETED],
      criterias: [constants.criteria.FILES_DELETED],
      criteriasFunctions: [STANDARD_HARD_CRITERIA_FUNCTION],
      description: `Delete ${constants.criteria.FILES_DELETED} files`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const filesMovedTemplate = (): StackingAchievementTemplate => ({
      name: 'Please bro just one more refactor %d',
      iconDir: 'mover',
      category: constants.category.FILES,
      group: 'Files Moved',
      labels: [constants.category.FILES, constants.labels.FILES_MOVED],
      criterias: [constants.criteria.FILES_MOVED],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Move (or rename) ${constants.criteria.FILES_MOVED} files`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });
  }

  export namespace git {
    //////////////////////// GIT STACKING TEMPLATES ////////////////////////
    export const commitsTemplate = (): StackingAchievementTemplate => ({
      name: 'Committer %d',
      iconDir: 'committer',
      category: constants.category.GIT,
      group: 'Commits',
      labels: [constants.category.GIT, constants.labels.COMMITS],
      criterias: [constants.criteria.COMMITS],
      criteriasFunctions: [STANDARD_HARD_CRITERIA_FUNCTION],
      description: `Commit ${constants.criteria.COMMITS} times`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const branchesCreatedTemplate = (): StackingAchievementTemplate => ({
      name: 'Friend of the Trees %d',
      iconDir: 'brancher',
      category: constants.category.GIT,
      group: 'Branches Created',
      labels: [constants.category.GIT, constants.labels.BRANCHES_CREATED],
      criterias: [constants.criteria.BRANCHES_CREATED],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Create ${constants.criteria.BRANCHES_CREATED} branches`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const mergesAndRebasesTemplate = (): StackingAchievementTemplate => ({
      name: 'Merger %d',
      iconDir: 'merger',
      category: constants.category.GIT,
      group: 'Merges and Rebases',
      labels: [constants.category.GIT, constants.labels.MERGES_AND_REBASES],
      criterias: [constants.criteria.MERGES_AND_REBASES],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Merge or rebase ${constants.criteria.MERGES_AND_REBASES} times`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const amendsTemplate = (): StackingAchievementTemplate => ({
      name: 'Amender %d',
      iconDir: 'amender',
      category: constants.category.GIT,
      group: 'Amends',
      labels: [constants.category.GIT, constants.labels.AMENDS],
      criterias: [constants.criteria.AMENDS],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Amend ${constants.criteria.AMENDS} times`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const forcedPushesTemplate = (): StackingAchievementTemplate => ({
      name: 'Good Luck Everyone %d',
      iconDir: 'forcer',
      category: constants.category.GIT,
      group: 'Forced Pushes',
      labels: [constants.category.GIT, constants.labels.FORCED_PUSHES],
      criterias: [constants.criteria.FORCED_PUSHES],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Force push ${constants.criteria.FORCED_PUSHES} times`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const pushesTemplate = (): StackingAchievementTemplate => ({
      name: 'Ship fast, talk later %d',
      iconDir: 'pusher',
      category: constants.category.GIT,
      group: 'Pushes',
      labels: [constants.category.GIT, constants.labels.PUSHES],
      criterias: [constants.criteria.PUSHES],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Push ${constants.criteria.PUSHES} times`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });
  }

  export namespace productivity {
    //////////////////////// PRODUCTIVITY STACKING TEMPLATES ////////////////////////
    export const linesOfCodeTemplate = (): StackingAchievementTemplate => ({
      name: 'Code Monkey %d',
      iconDir: 'code_monkey',
      category: constants.category.PRODUCTIVITY,
      group: 'Lines of Code',
      labels: [constants.category.PRODUCTIVITY, constants.labels.LINES_OF_CODE],
      criterias: [constants.criteria.LINES_OF_CODE],
      criteriasFunctions: [STANDARD_EASY_CRITERIA_FUNCTION],
      description: `Write ${constants.criteria.LINES_OF_CODE} lines of code`,
      minTier: 0,
      maxTier: 10,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const linesOfCodeLanguageTemplates = (): StackingAchievementTemplate[] =>
      constants.labels.LANGUAGES.map(language => ({
        name: `LANGUAGE Speaker %d`.replace('LANGUAGE', language),
        iconDir: `%s_speaker`.replace('%s', language),
        category: constants.category.PRODUCTIVITY,
        group: 'Lines of Code',
        labels: [constants.category.PRODUCTIVITY, constants.labels.LINES_OF_CODE],
        criterias: [constants.criteria.LINES_OF_CODE_LANGUAGE.replace('%s', language)],
        criteriasFunctions: [STANDARD_MEDIUM_CRITERIA_FUNCTION],
        description: `Write LOC lines of LANGUAGE code`.replace('LANGUAGE', language).replace('LOC', constants.criteria.LINES_OF_CODE_LANGUAGE.replace('%s', language)),
        minTier: 0,
        maxTier: 15,
        pointsFunction: STANDARD_POINTS_FUNCTION,
        hidden: false,
        requires: [],
      }));

    export const linesOfCommentTemplate = (): StackingAchievementTemplate => ({
      name: 'Outstanding Commentator %d',
      iconDir: 'commentator',
      category: constants.category.PRODUCTIVITY,
      group: 'Lines of Comments',
      labels: [constants.category.PRODUCTIVITY, constants.labels.LINES_OF_COMMENTS],
      criterias: [constants.criteria.LINES_OF_COMMENTS],
      criteriasFunctions: [STANDARD_MEDIUM_CRITERIA_FUNCTION],
      description: `Write ${constants.criteria.LINES_OF_COMMENTS} lines of comments`,
      minTier: 0,
      maxTier: 15,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const numberOfPastesTemplate = (): StackingAchievementTemplate => ({
      name: 'Copy Ninja %d',
      iconDir: 'copy_ninja',
      category: constants.category.PRODUCTIVITY,
      group: 'Number of Pastes',
      labels: [constants.category.PRODUCTIVITY, constants.labels.NUMBER_OF_PASTES],
      criterias: [constants.criteria.NUMBER_OF_PASTES],
      criteriasFunctions: [STANDARD_HARD_CRITERIA_FUNCTION],
      description: `Paste ${constants.criteria.NUMBER_OF_PASTES} times`,
      minTier: 0,
      maxTier: 15,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const errorsFixedTemplate = (): StackingAchievementTemplate => ({
      name: 'I can fix her %d',
      iconDir: 'error_fixer',
      category: constants.category.PRODUCTIVITY,
      group: 'Errors Fixed',
      labels: [constants.category.PRODUCTIVITY, constants.labels.ERRORS_FIXED],
      criterias: [constants.criteria.ERRORS_FIXED],
      criteriasFunctions: [STANDARD_HARD_CRITERIA_FUNCTION],
      description: `Fix ${constants.criteria.ERRORS_FIXED} errors`,
      minTier: 0,
      maxTier: 15,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const timeSpentTemplate = (): StackingAchievementTemplate => ({
      name: 'Shower Avoider %d',
      iconDir: 'shower_avoider',
      category: constants.category.PRODUCTIVITY,
      group: 'Time Spent',
      labels: [constants.category.PRODUCTIVITY, constants.labels.TIME_SPENT],
      criterias: [constants.criteria.TIME_SPENT],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Spend ${constants.criteria.TIME_SPENT} hours coding`,
      minTier: 0,
      maxTier: 15,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const numberOfSimultaneousTabsTemplate = (): StackingAchievementTemplate => ({
      name: 'Tab Hoarder %d',
      iconDir: 'tab_hoarder',
      category: constants.category.PRODUCTIVITY,
      group: 'Number of Simultaneous Tabs',
      labels: [constants.category.PRODUCTIVITY, constants.labels.NUMBER_OF_SIMULTANEOUS_TABS],
      criterias: [constants.criteria.NUMBER_OF_SIMULTANEOUS_TABS],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Open ${constants.criteria.NUMBER_OF_SIMULTANEOUS_TABS} tabs simultaneously`,
      minTier: 0,
      maxTier: 5,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });
  }

  export namespace vscode {
    //////////////////////// VSCODE STACKING TEMPLATES ////////////////////////
    export const extensionsInstalledTemplate = (): StackingAchievementTemplate => ({
      name: 'Got to catch them all ! %d',
      iconDir: 'extension_master',
      category: constants.category.VSCODE,
      group: 'Extensions Installed',
      labels: [constants.category.VSCODE, constants.labels.EXTENSIONS_INSTALLED],
      criterias: ['extensionsInstalledCount'],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Install ${'extensionsInstalledCount'} extensions`,
      minTier: 0,
      maxTier: 5,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const extensionsOutdatedTemplate = (): StackingAchievementTemplate => ({
      name: 'Living Dangerously %d',
      iconDir: 'extension_outdated',
      category: constants.category.VSCODE,
      group: 'Extensions Outdated',
      labels: [constants.category.VSCODE, constants.labels.EXTENSIONS_OUTDATED],
      criterias: ['extensionsOutdatedCount'],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Have ${'extensionsOutdatedCount'} outdated extensions`,
      minTier: 0,
      maxTier: 5,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });

    export const themesInstalledTemplate = (): StackingAchievementTemplate => ({
      name: 'Pimp my Ride %d',
      iconDir: 'theme_master',
      category: constants.category.VSCODE,
      group: 'Themes Installed',
      labels: [constants.category.VSCODE, constants.labels.THEMES_INSTALLED],
      criterias: ['themesInstalledCount'],
      criteriasFunctions: [STANDARD_INFERNAL_CRITERIA_FUNCTION],
      description: `Install ${'themesInstalledCount'} themes`,
      minTier: 0,
      maxTier: 5,
      pointsFunction: STANDARD_POINTS_FUNCTION,
      hidden: false,
      requires: [],
    });
  }
}

File: achievements/StackingTemplates.py
'''
StackingTemplates module contains the templates definitions for the achievements

List of Stacking Tempaltes:
- Files
    - Files Created
    - Files Created for each language
    - Files Deleted
    - Files Moved
- Git
    - Commits
    - Branches created
    - Merges and rebases
    - Amends
    - Forced Pushes
    - Pushes
- Productivity
    - Number of sumultaneous tabs
    - Time Spent
    - Errors Fixed
    - Lines of Code
    - Lines of Code for each language
    - Lines of comments
    - Number of pastes
- VSCode
    - Extensions Installed
    - Extensions Outdated
    - Themes Installed
'''

from typing import List, Callable
import constants

class StackingTemplates:
    '''
    Stacking templates class, contains the stacking templates for the achievements
    '''

    STANDARD_EASY_CRITERIA_FUNCTION = lambda x: 10**x                                                                            # [1, 10, 100, 1000, 10000, 100000,...]
    STANDARD_MEDIUM_CRITERIA_FUNCTION = lambda x: 2 ** ((x + 2) // 2) * 5 ** ((x + 1) // 2)                                      # [2, 10, 20, 100, 200, 1000, 2000, 10000, 20000, 100000]
    STANDARD_HARD_CRITERIA_FUNCTION = lambda x: 2 ** (x // 2) * 5 ** ((x + 1) // 2)                                              # [1, 5, 10, 50, 100, 500, 1000, 5000, 10000, 50000,...]
    STANDARD_INFERNAL_CRITERIA_FUNCTION = lambda x: (2 ** max(0, (x - 1) // 2)) * (3 ** (1 if x == 1 else 0)) * (5 ** (x // 2))  # [1, 3, 5, 10, 50, 100, 500, 1000, 5000, 10000,...]

    STANDARD_POINTS_FUNCTION = STANDARD_INFERNAL_CRITERIA_FUNCTION

    class Files:
        '''
        Files stacking templates
        '''
        @staticmethod
        def filesCreatedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the files created achievement
            '''
            name = 'Creator %d'
            icon_dir = 'creator'
            category = constants.Category.FILES
            group = 'Files Created'
            criterias = [constants.Criteria.FILES_CREATED]
            description = f'Create {criterias[0]} files'
            labels = [constants.Category.FILES, constants.Labels.FILES_CREATED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_MEDIUM_CRITERIA_FUNCTION], description, 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def filesCreatedLanguageTemplates() -> List[tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]]:
            '''
            Returns the templates for the files created for each language achievement
            '''
            name = 'LANGUAGE Creator %d'
            icon_dir = '%s_creator'
            category = constants.Category.FILES
            group = 'Files Created'
            description = 'Create FILES files in LANGUAGE'
            labels = [constants.Category.FILES, constants.Labels.FILES_CREATED]

            languages_tuples = []
            for language in constants.Labels.LANGUAGES:
                criterias = [constants.Criteria.FILES_CREATED_LANGUAGE % language]
                languages_tuples.append((name.replace('LANGUAGE', language), icon_dir % language, category, group, labels, criterias, [StackingTemplates.STANDARD_HARD_CRITERIA_FUNCTION], description.replace('LANGUAGE', language).replace('FILES', criterias[0]), 0, 11, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []))

            return languages_tuples


        @staticmethod
        def filesDeletedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the files deleted achievement
            '''
            name = 'Deleter %d'
            icon_dir = 'deleter'
            category = constants.Category.FILES
            group = 'Files Deleted'
            criterias = [constants.Criteria.FILES_DELETED]
            description = f'Delete {criterias[0]} files'
            labels = [constants.Category.FILES, constants.Labels.FILES_DELETED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_HARD_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def filesMovedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the files moved achievement
            '''
            name = 'Please bro just one more refactor %d'
            icon_dir = 'mover'
            category = constants.Category.FILES
            group = 'Files Moved'
            criterias = [constants.Criteria.FILES_MOVED]
            description = f'Move (or rename) {criterias[0]} files'
            labels = [constants.Category.FILES, constants.Labels.FILES_MOVED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


    class Git:
        '''
        Git stacking templates
        '''
        @staticmethod
        def commitsTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the commits achievement
            '''
            name = 'Committer %d'
            icon_dir = 'committer'
            category = constants.Category.GIT
            group = 'Commits'
            criterias = [constants.Criteria.COMMITS]
            description = f'Commit {criterias[0]} times'
            labels = [constants.Category.GIT, constants.Labels.COMMITS]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_HARD_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def branchesCreatedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the branches created achievement
            '''
            name = 'Friend of the Trees %d'
            icon_dir = 'brancher'
            category = constants.Category.GIT
            group = 'Branches Created'
            criterias = [constants.Criteria.BRANCHES_CREATED]
            description = f'Create {criterias[0]} branches'
            labels = [constants.Category.GIT, constants.Labels.BRANCHES_CREATED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def mergesAndRebasesTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the merges and rebases achievement
            '''
            name = 'Merger %d'
            icon_dir = 'merger'
            category = constants.Category.GIT
            group = 'Merges and Rebases'
            criterias = [constants.Criteria.MERGES_AND_REBASES]
            description = f'Merge or rebase {criterias[0]} times'
            labels = [constants.Category.GIT, constants.Labels.MERGES_AND_REBASES]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def amendsTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the amends achievement
            '''
            name = 'Amender %d'
            icon_dir = 'amender'
            category = constants.Category.GIT
            group = 'Amends'
            criterias = [constants.Criteria.AMENDS]
            description = f'Amend {criterias[0]} times'
            labels = [constants.Category.GIT, constants.Labels.AMENDS]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def forcedPushesTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the forced pushes achievement
            '''
            name = 'Good Luck Everyone %d'
            icon_dir = 'forcer'
            category = constants.Category.GIT
            group = 'Forced Pushes'
            criterias = [constants.Criteria.FORCED_PUSHES]
            description = f'Force push {criterias[0]} times'
            labels = [constants.Category.GIT, constants.Labels.FORCED_PUSHES]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def pushesTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the pushes achievement
            '''
            name = 'Ship fast, talk later %d'
            icon_dir = 'pusher'
            category = constants.Category.GIT
            group = 'Pushes'
            criterias = [constants.Criteria.PUSHES]
            description = f'Push {criterias[0]} times'
            labels = [constants.Category.GIT, constants.Labels.PUSHES]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


    class Productivity:
        '''
        Productivity stacking templates
        '''
        @staticmethod
        def linesOfCodeTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the lines of code achievement
            '''
            name = 'Code Monkey %d'
            icon_dir = 'code_monkey'
            category = constants.Category.PRODUCTIVITY
            group = 'Lines of Code'
            criterias = [constants.Criteria.LINES_OF_CODE]
            description = f'Write {criterias[0]} lines of code'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.LINES_OF_CODE]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_EASY_CRITERIA_FUNCTION], description, 0, 10, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def linesOfCodeLanguageTemplates() -> List[tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]]:
            '''
            Returns the templates for the lines of code for each language achievement
            '''
            name = 'LANGUAGE Speaker %d'
            icon_dir = '%s_speaker'
            category = constants.Category.PRODUCTIVITY
            group = 'Lines of Code'
            description = 'Write LOC lines of LANGUAGE code'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.LINES_OF_CODE]

            languages_tuples = []
            for language in constants.Labels.LANGUAGES:
                criterias = [constants.Criteria.LINES_OF_CODE_LANGUAGE % language]
                languages_tuples.append((name.replace('LANGUAGE', language), icon_dir % language, category, group, labels, criterias, [StackingTemplates.STANDARD_MEDIUM_CRITERIA_FUNCTION], description.replace('LANGUAGE', language).replace('LOC', criterias[0]), 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []))

            return languages_tuples


        @staticmethod
        def linesOfCommentTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the lines of comments achievement
            '''
            name = 'Outstanding Commentator %d'
            icon_dir = 'commentator'
            category = constants.Category.PRODUCTIVITY
            group = 'Lines of Comments'
            criterias = [constants.Criteria.LINES_OF_COMMENTS]
            description = f'Write {criterias[0]} lines of comments'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.LINES_OF_COMMENTS]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_MEDIUM_CRITERIA_FUNCTION],  description, 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def numberOfPastesTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the number of pastes achievement
            '''
            name = 'Copy Ninja %d'
            icon_dir = 'copy_ninja'
            category = constants.Category.PRODUCTIVITY
            group = 'Number of Pastes'
            criterias = [constants.Criteria.NUMBER_OF_PASTES]
            description = f'Paste {criterias[0]} times'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.NUMBER_OF_PASTES]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_HARD_CRITERIA_FUNCTION], description, 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def errorsFixedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the errors fixed achievement
            '''
            name = 'I can fix her %d'
            icon_dir = 'error_fixer'
            category = constants.Category.PRODUCTIVITY
            group = 'Errors Fixed'
            criterias = [constants.Criteria.ERRORS_FIXED]
            description = f'Fix {criterias[0]} errors'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.ERRORS_FIXED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_HARD_CRITERIA_FUNCTION], description, 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []

        @staticmethod
        def timeSpentTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the time spent achievement
            '''
            name = 'Shower Avoider %d'
            icon_dir = 'shower_avoider'
            category = constants.Category.PRODUCTIVITY
            group = 'Time Spent'
            criterias = [constants.Criteria.TIME_SPENT]
            description = f'Spend {criterias[0]} hours coding'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.TIME_SPENT]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 15, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def numberOfSimultaneousTabsTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the number of simultaneous tabs achievement
            '''
            name = 'Tab Hoarder %d'
            icon_dir = 'tab_hoarder'
            category = constants.Category.PRODUCTIVITY
            group = 'Number of Simultaneous Tabs'
            criterias = [constants.Criteria.NUMBER_OF_SIMULTANEOUS_TABS]
            description = f'Open {criterias[0]} tabs simultaneously'
            labels = [constants.Category.PRODUCTIVITY, constants.Labels.NUMBER_OF_SIMULTANEOUS_TABS]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 5, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []

    class VSCode:
        '''
        VSCode stacking templates
        '''

        @staticmethod
        def extensionsInstalledTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the extensions installed achievement
            '''
            name = 'Got to catch them all ! %d'
            icon_dir = 'extension_master'
            category = constants.Category.VSCODE
            group = 'Extensions Installed'
            criterias = ['extensionsInstalledCount']
            description = f'Install {criterias[0]} extensions'
            labels = [constants.Category.VSCODE, constants.Labels.EXTENSIONS_INSTALLED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 5, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def extensionsOutdatedTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the extensions outdated achievement
            '''
            name = 'Living Dangerously %d'
            icon_dir = 'extension_outdated'
            category = constants.Category.VSCODE
            group = 'Extensions Outdated'
            criterias = ['extensionsOutdatedCount']
            description = f'Have {criterias[0]} outdated extensions'
            labels = [constants.Category.VSCODE, constants.Labels.EXTENSIONS_OUTDATED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 5, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []


        @staticmethod
        def themesInstalledTemplate() -> tuple[str, str, str, str, List[str], List[str], List[Callable], str, int, int, Callable, bool, List[int]]:
            '''
            Returns the template for the themes installed achievement
            '''
            name = 'Pimp my Ride %d'
            icon_dir = 'theme_master'
            category = constants.Category.VSCODE
            group = 'Themes Installed'
            criterias = ['themesInstalledCount']
            description = f'Install {criterias[0]} themes'
            labels = [constants.Category.VSCODE, constants.Labels.THEMES_INSTALLED]

            return name, icon_dir, category, group, labels, criterias, [StackingTemplates.STANDARD_INFERNAL_CRITERIA_FUNCTION], description, 0, 5, StackingTemplates.STANDARD_POINTS_FUNCTION, False, []
